x/btcstaking/keeper/finality_providers.go

```diff
@@ -67,6 +68,35 @@ func (k Keeper) SlashFinalityProvider(ctx context.Context, fpBTCPK []byte) error
 	return nil
 }
 
+func (k Keeper) PropagateFPSlashingToConsumers(ctx context.Context, fpBTCPK *bbn.BIP340PubKey) error {
+	// Get all delegations for this finality provider
+	delegations, err := k.getFPBTCDelegations(ctx, fpBTCPK)
+	if err != nil {
+		return err
+	}
+
+	for _, delegation := range delegations {
+		// Create SlashedBTCDelegation event
+		consumerEvent := types.CreateSlashedBTCDelegationEvent(delegation)
+
+		// Get consumer IDs of non-Babylon finality providers
+		restakedFPConsumerIDs, err := k.restakedFPConsumerIDs(ctx, delegation.FpBtcPkList)
+		if err != nil {
+			return err
+		}
+
+		// Send event to each involved consumer chain
+		for _, consumerID := range restakedFPConsumerIDs {
+			// TODO: repeated set ops in KV store is inefficient; consider refactoring
+			if err := k.AddBTCStakingConsumerEvent(ctx, consumerID, consumerEvent); err != nil {
```

Comment:
As the first step, we can refactor the function to take a list of `consumerEvent`s, append them to the `BTCStakingIBCPacket`, and set it back to DB